asyncpg/connect_utils.py

```diff
@@ -435,7 +487,12 @@ def _parse_connect_arguments(*, dsn, host, port, user, password, passfile,
     raise last_error
 
 
-async def _get_ssl_ready_socket(host, port, *, loop):
+async def _use_ssl_ready_socket(host, port, partial, *, loop, ssl,
+                                server_hostname, ssl_is_advisory=False):
+    # Note: ssl_is_advisory only affects behavior when the server does not
+    # accept SSLRequests. If the SSLRequest is accepted but either the SSL
+    # negotiation fails or the PostgreSQL user isn't permitted to use SSL,
+    # there's nothing that would attempt to reconnect with a non-SSL socket.
```

Comment:
I'm pretty sure the libpq behavior is to retry in at least some of these cases that this implementation here does not. So, sslmode='prefer' probably still isn't complete.

...but I guess given that sslmode='prefer' implies no cert checking, SSL negotiation failure seems a bit of an edge case. And I assume that pg_hba.confs banning SSL when non-SSL would have worked are uncommon too. So, maybe we're fine for the majority use-cases of sslmode='prefer' anyways?

Reply:
In my opinion, sslmode='prefer' is a legacy, dangerous vestige, which should be actively discouraged.  You either have proper TLS (with or without host verification), or you don't.  For this reason I don't think adding the `ssl_is_advisory` parameter to `connect()` is a good idea.  Let's restrict the ability to choose the downgradable SSL mode exclusively to DSN, and only because libpq still supports this and some existing environments might still rely on this behavior.